llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp

```diff
@@ -770,6 +770,15 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Instruction *I,
             return InsertNewInstWith(Shl, I->getIterator());
           }
         }
+
+        const APInt *Factor;
+        if (match(I->getOperand(0),
+                  m_OneUse(m_Mul(m_Value(X), m_APInt(Factor)))) &&
+            Factor->countr_zero() >= ShiftAmt) {
+          BinaryOperator *Mul = BinaryOperator::CreateMul(
```

Comment:
You can preserve `nuw` and `nsw` iff you have `nuw`:
https://alive2.llvm.org/ce/z/P9NudG

Reply:
`(X *nuw (Y << K)) u>> K` has been handled in other places: https://godbolt.org/z/43W4vE8bK

Reply:
Can we put the two impls in the same place?

Reply:
I don't think the two patterns would combine well. In one case we can optimize because there is no overflow. In the other case we can optimize because the bits affected by overflow are not demanded.